test/test_type_hints.py

```diff
@@ -170,6 +170,23 @@ def test_doc_examples(self):
             except subprocess.CalledProcessError as e:
                 raise AssertionError("mypy failed.  Look above this error for mypy's output.")
 
+    @unittest.skipIf(sys.version_info[0] == 2, "no type hints for Python 2")
+    @unittest.skipIf(not HAVE_MYPY, "need mypy")
+    def test_type_hint_examples(self):
+        """
+        Runs mypy over all the test examples present in
+        type_hint_tests directory
+        """
+    test_path = os.path.dirname(os.path.realpath(__file__))
+    examples_folder = os.path.join(test_path, "type_hint_tests")
+    examples = os.listdir(examples_folder)
+    for i in examples:
+        example_path = os.path.join(os.path.dirname(os.path.realpath(__file__)),
+                                    "type_hint_tests", i)
+        process = subprocess.Popen(['mypy', example_path], stdout=subprocess.PIPE)
+        (stdout, stderr) = process.communicate()
+        if stdout:
+            raise AssertionError('Type hint tests not passed in ', i)
```

Comment:
The tests on macos are failing with:
```
Mar 12 07:43:31 Traceback (most recent call last):
Mar 12 07:43:31   File "test_type_hints.py", line 112, in <module>
Mar 12 07:43:31     class TestTypeHints(TestCase):
Mar 12 07:43:31   File "test_type_hints.py", line 186, in TestTypeHints
Mar 12 07:43:31     process = subprocess.Popen(['mypy', example_path], stdout=subprocess.PIPE)
Mar 12 07:43:31   File "/Users/distiller/workspace/miniconda3/lib/python3.7/subprocess.py", line 800, in __init__
Mar 12 07:43:31     restore_signals, start_new_session)
Mar 12 07:43:31   File "/Users/distiller/workspace/miniconda3/lib/python3.7/subprocess.py", line 1551, in _execute_child
Mar 12 07:43:31     raise child_exception_type(errno_num, err_msg, err_filename)
Mar 12 07:43:31 FileNotFoundError: [Errno 2] No such file or directory: 'mypy': 'mypy'
```
(see https://circleci.com/gh/pytorch/pytorch/4781861?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)

I suspect that mypy is set up differently on our mac os CI machines. Could we change the `subprocess.Popen` call to call mypy like how the other test (test_doc_examples) in this file does? 

So we'd replace these lines with something like the following:

```
for example in examples:
    try: 
        example_path = os.path.join(examples_folder, example)
        subprocess.run([ 
            sys.executable, 
            '-mmypy', 
            '--follow-imports', 'silent', 
            '--check-untyped-defs', 
            example_path], 
            cwd=tmp_dir, 
            check=True) 
    except subprocess.CalledProcessError as e: 
        raise AssertionError("mypy failed for example {}.  Look above this error for mypy's output.".format(example)) 
```

Reply:
After this changes, I'm getting this
```
Fail to import hypothesis in common_utils, tests are not derandomized
test/type_hint_tests/size.py:7: error: Tuple index out of range
test/type_hint_tests/size.py:8: error: Tuple index out of range
Traceback (most recent call last):
  File "test/test_type_hints.py", line 194, in TestTypeHints
    check=True) 
  File "/usr/lib/python3.6/subprocess.py", line 438, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['/usr/bin/python3', '-mmypy', '--follow-imports', 'silent', '--check-untyped-defs', '/home/prajjwal/Desktop/Github/pytorch/test/type_hint_tests/size.py']' returned non-zero exit status 1.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test/test_type_hints.py", line 112, in <module>
    class TestTypeHints(TestCase):
  File "test/test_type_hints.py", line 196, in TestTypeHints
    raise AssertionError("mypy failed for example {}.  Look above this error for mypy's output.".format(example))
AssertionError: mypy failed for example size.py.  Look above this error for mypy's output.
```
I wanted to keep error message short, that's why I was using `subprocess.Popen` as it provides more control I guess (not sure). Do you think this is fine ?

Reply:
If this is fine, I can push this change now.

Reply:
Yeah, I think this is fine. The other thing about the current PR is that we don't print out exactly what is wrong to the user and this would also solve that problem. I like to err on the side of being more verbose.

Reply:
Please check now.

Reply:
It looks great now!

Reply:
The error message complies with `TestCase` class now. For example:
```
Fail to import hypothesis in common_utils, tests are not derandomized
.test/type_hint_tests/size.py:7: error: Tuple index out of range
test/type_hint_tests/size.py:8: error: Tuple index out of range
F
======================================================================
FAIL: test_type_hint_examples (__main__.TestTypeHints)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_type_hints.py", line 194, in test_type_hint_examples
    check=True)
subprocess.CalledProcessError: Command '['/usr/bin/python3', '-mmypy', '--follow-imports', 'silent', '--check-untyped-defs', '/home/prajjwal/Desktop/Github/pytorch/test/type_hint_tests/size.py']' returned non-zero exit status 1.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test/test_type_hints.py", line 196, in test_type_hint_examples
    raise AssertionError("mypy failed for example {}.  Look above this error for mypy's output.".format(example))
AssertionError: mypy failed for example size.py.  Look above this error for mypy's output.

----------------------------------------------------------------------
Ran 2 tests in 20.238s

FAILED (failures=1)

```

Reply:
Please recheck now @zou3519 . All the tests are passing (including linting). Let me know if more changes are required.